### PR TITLE
Add automatic updater

### LINKED ISSUE #95: Automatic Updater Script
I was thinking about how a user could update to a newer version of the uploader and it seems like it would require a lot of manual work, such as deleting all of the old project files but keeping their config and uploads, then selectively copying the code files for the new version. What if instead, this was done through some sort of automatic updater script? For example, since releases now come packaged in a folder called `release`, all a user would need to do to update their uploader would be to copy this entire release folder into their current uploader. Then, the next time the user opened the gallery page, the uploader would check if there's a folder called `release` present, and if there is, it would prompt to update to a new version of the uploader. If the user says yes, it would open a PHP page called `updater.php` located *in the new `release` folder*. The update script would then copy files from the release folder and overwrite them in the main uploader, but only for non-user configurable files. Then, all the user would have to do would be to delete the release folder.

For example, if a user set up their uploader in a folder called `/my/website/root/`:
1. The user copies the `release` folder from the new version into `/my/website/root/release/`
2. The user opens up their gallery page at `http://mywebsite.com` and it shows an Update button
3. The user presses update, which sends the user to `http://mywebsite.com/release/updater.php`
4. This script overwrites the project code in `/my/website/root/` with the new code from `/my/website/root/release/` but keeps the old config files
5. The user deletes the `/my/website/root/release` folder

This would hopefully provide a more streamlined update process. Let me know what you think. I'd be happy to work on this.

### PR DESCRIPTION
**Pull request by theaquarium**
Resolves #95 

This adds automatic updating functionality exactly as described in the linked issue. All a user has to do to update is copy the `release` folder to their site, click a button, and voila! However, this script doesn't modify user configuration, so it shouldn't break any installs. That also means that users will need to manually edit the configuration if the config schema changes in the future.

### CODE REVIEW

**Comment by theaquarium** [comment]
Ok, so now I've rewritten the updater so that you don't need to manually download the new update anymore. The uploader now will check GitHub for new versions and if it detects one, it will prompt to update. If the user agrees, the new update is downloaded, unzipped, and installed.

Users have the option of completely disabling the updater, as well as making a backup of the updated files. The backup option is on by default and just copies the old versions of files that get updated into another folder so that users can easily roll back to an older version. 

Users will still need to manually update any configuration schema changes, but the uploader asks users to *please read the changelog* twice before the update starts.

This is meant to really streamline the update process so that users really don't have to think about updates and so that we can push crucial updates if we need to, but if users want to run their uploader without being affected by any changes we make, they can just turn updates off.

Let me know what you think.

**Comment by davwheat** [comment]
I think that we should add a form of Migration system, whete, in the case of config schema changes, we can run a set of migrations which update the config. The issue is that it would be difficult to parse the config.

**Comment by theaquarium** [comment]
That's fair. This would be pretty easy to add if config options are added, but not if any are removed. Maybe it could work by looping through the new config and checking if each option exists in the old config and if it doesn't copying the value (which would be the default value) over. Then, users could look through and see what they need to change. Writing the config back into the file would be extremely challenging though. I think the best way to go around this is to just use JSON config, but it might be too late for that.

**Comment by theaquarium** [comment]
Another proposition, instead of actually changing the config, we can have a default config file which is used whenever a config option is unavailable. This way, we don't touch the user's config at all and they can go and change their config on their own later. To help with this, maybe the uploader could show a little banner which lists which options the default config is being used for so that the users know what they need to add.

**Comment by theaquarium** [comment]
Alright, I've implemented it basically as I described in the second comment. All files now load config from `merge_config.php` instead of `config.php`. The `merge_config.php` file combines user config with default config and makes note of all missing options. Then, the list of missing options is displayed on the gallery page like so:

image

I think this is a pretty good way to notify users of errors like this without having to touch their config. Let me know what you think.

File: src/error.php [hunk 1]
```diff
@@ -11,25 +11,23 @@
...
+    <link rel="stylesheet" href="css/main.css">
     <title>Error - <?php echo $config['page_title']; ?></title>
     
-    <link rel="apple-touch-icon" sizes="180x180" href="<?=$config['upload_access_path']?>icons/apple-touch-icon.png">
-    <link rel="icon" type="image/png" sizes="32x32" href="<?=$config['upload_access_path']?>icons/favicon-32x32.png">
-    <link rel="icon" type="image/png" sizes="16x16" href="<?=$config['upload_access_path']?>icons/favicon-16x16.png">
-    <link rel="manifest" href="<?=$config['upload_access_path']?>manifest.webmanifest">
-    <link rel="mask-icon" href="<?=$config['upload_access_path']?>icons/safari-pinned-tab.svg" color="#5c5cbc">
-    <link rel="shortcut icon" href="<?=$config['upload_access_path']?>favicon.ico">
+    <link rel="apple-touch-icon" sizes="180x180" href="icons/apple-touch-icon.png">
```
Comment (JoeGandy) [inline src/error.php:22]: You've kind of merged your two branches together, once you've merged in #106 I'll come back and review this :)

Comment (theaquarium) [reply src/error.php:22]: Haha you're right, I must have not branched off from master.

**Comment by theaquarium** [comment]
There, that should have fixed it.

**Review by JoeGandy — APPROVED** [review]

**Comment by JoeGandy** [comment]
Sorry for the delay in approving! Looks great

**Pull request merged**